nifi-stateless/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/flow/StandardStatelessFlowCurrent.java

```diff
@@ -57,23 +58,22 @@ private StandardStatelessFlowCurrent(final Builder builder) {
         this.processContextFactory = builder.processContextFactory;
     }
 
-    public Connectable getCurrentComponent() {
-        return currentComponent;
-    }
-
     @Override
     public void triggerFlow() {
         try {
             boolean completionReached = false;
             while (!completionReached) {
+                tracker.shelveReadyComponents();
```

Comment:
Is shelving the previously ready components really needed?

When I found the issue, I quickly fixed it with using the existing `ready` `LinkedHashSet` and implementing a simple `getNextReady()` method in `AsynchronousCommitTracker` which returns the top element of the ready "stack". I tested it again with several flows and could not find any issues with that. The advantage would be that the code is simpler and more straightforward (less datastructures, less loops).

At this point in the execution, the ready stack contains the component that triggered the root components for more input (at the top of the stack), and may contain some other already ready components. During the re-execution of the flow from the root components, these components may become ready again and they will be moved to the top of the stack or may just be waiting in the stack in their original position (as they were shelved). In both cases they will be triggered when their turn comes and will process all the FlowFiles from their input queue. That's why I think that shelving is not needed.

The code would look like this:
```
            while (!completionReached) {
                triggerRootConnectables();

                while (tracker.isAnyReady()) {
                    final Connectable connectable = tracker.getNextReady();
                    logger.debug("The next ready component to be triggered: {}", connectable);

                    // Continually trigger the given component as long as it is ready to be triggered
                    NextConnectable nextConnectable = triggerWhileReady(connectable);

                    // If there's nothing left to do, return
                    if (nextConnectable == NextConnectable.NONE) {
                        return;
                    }

                    // If next connectable is whatever is ready, just continue loop
                    if (nextConnectable == NextConnectable.NEXT_READY) {
                        continue;
                    }

                    // Otherwise, we need to break out of this loop so that we can trigger root connectables or complete dataflow
                    break;
                }

                // We have reached completion if the tracker does not know of any components ready to be triggered AND
                // we have no data queued in the flow (with the exception of Output Ports).
                completionReached = !tracker.isAnyReady() && isFlowQueueEmpty();
            }
```

Reply:
@turcsanyip can you share the proposed code for a getNextReady() method for the `AsynchronousCommitTracker`?

Reply:
It returns the last element from the ready list (unfortunately there is no "getLastElement()" in `LinkedHashSet`):
```
    public Connectable getNextReady() {
        if (ready.isEmpty()) {
            return null;
        }
        return new ArrayList<>(ready).get(ready.size() - 1);
    }
```

Reply:
I see. The key here being that in this approach the getNextReady() would return the last element in the list, rather than the List of all ready components. I was missing that distinction at first. I think you're right - this should work essentially the same as the solution that I put in this PR, but your method is cleaner. Simpler. And possibly more efficient, as well.
Will update the PR to go this route. Thanks! I'll probably avoid creating the new ArrayList and returning the last element, though, and instead just iterate over the Set to get the last element - just to avoid that intermediate ArrayList. But otherwise I think your solution is really good. Nice job!